crates/ra_mbe/src/syntax_bridge.rs

```diff
@@ -116,6 +117,42 @@ impl TokenMap {
     }
 }
 
+impl RevTokenMap {
+    fn add(&mut self, relative_range: TextRange, token_id: tt::TokenId) {
+        self.ranges.push((relative_range, token_id.clone()))
+    }
+
+    /// Map a given token map to (Expanded syntax node, Input tokens) text-ranges pair
+    ///
+    /// This function do the following things:
+    ///
+    /// 1. Undo the increment of token-id `shift`:
+    ///     When we output a token from from macro argument, we increased its id
+    ///     by `shift` (so it's guaranteed to not to collide with anything from the definition)
+    ///     We undo the increment here to rollback to its original token id.
+    /// 2. Offset the input tokens (`to`) by `parent` text-range:
+    ///     We transforms the input tokens text-ranges from relative to original first token
+    ///     to parent text-range
+    /// 3. Maps expanded tokens text-ranges to parent text-ranges
+    ///
+    pub fn map_ranges(
+        &self,
+        to: &TokenMap,
+        parent: TextRange,
```

Comment:
Note that this function only uses `parent.start()`, so a slightly better API would have been to pass in a `TextUnit` here, instead of a `TextRange`.

Reply:
However, I think it might be better to maybe completely remove this function? 

We use it to precompute `Vec<TextRange, TextRange>` maps. However, as the lookup in those maps is linear anyway, I don't think that this pre-computation helps us. And it violates that philosophical idea of not storing derived data. 

Moreover, this function and `find_range` should really be understood together, and at the moment this is hard, because they live in different crates. I suggest moving this logic to `hir_expand`, and, instead of precomputing the map, do all this in the `find_range`. I expect it migth loook like this:

```rust
pub struct ExpansionInfo {
    arg_file: HirFileId,
    def_file: HirFileId,

    shift: u32,
    arg_map: TokenMap,
    def_map: TokenMap,
    exp_map: RevTokenMap,
}

impl ExpansionInfo {
    pub fn find_range(&self, from: TextRange) -> Option<(HirFileId, TextRange)> {
      let token_id = lookp_in_rev_map(&self.exp_map, from);
      let (token_map, file_id) = if token_id > self.shift { (&self.arg_map, self.arg_file) } else {  (&self.def_map, self.def_file)  };
      let range = lookup_in_map(token_map, token_id);
      Some((file_id, range + ...))
    }
}
```

Reply:
Fixed.